File: InterviewPrep/RecursiveDigitSum.cs
using System;
using System.Numerics;

static class Result
{
    public static int SuperDigit(string n, int k)
    {
        long start = 0;

        // First, we can iterate each character directly and add up their numeric
        // values directly. This allows us to not need the BigInteger class and
        // be a bit more efficient.

        foreach (char c in n)
        {
            start += (long) Char.GetNumericValue(c);
        }

        // Since multiplying the string 'k' times only means we will get the same
        // pattern again, we can just take the calculation of one iteration, and
        // multiply it 'k' times instead.

        long result = SuperDigitHelper(start * k);
        return (int) result;
    }

    private static long SuperDigitHelper(long num)
    {
        // If we only have one digit, then we're done.
        if (num < 10)
            return num;

        // If we still have more than one digit, then we have to calculate the
        // sum of its digits, and then iterate, or better said, recurse, on that
        // result. To do this, we calculate the % 10 of the number to get the
        // next digit, and then divide by 10 to "shift" to the next digit.

        long nextSuper = 0;

        while (num > 0)
        {
            nextSuper += num % 10;
            num /= 10;
        }

        return SuperDigitHelper(nextSuper);
    }
}

class Solution
{
    static int Main(string[] args)
    {
        string[] firstMultipleInput = Console.ReadLine().TrimEnd().Split(' ');
        string n = firstMultipleInput[0];

        int k = Convert.ToInt32(firstMultipleInput[1]);
        int result = Result.SuperDigit(n, k);

        Console.WriteLine(result);
        return 0;
    }
}

File: CompetitiveProgramming/Hackerrank/Medium/recursivedigitsum.mjs
#!/usr/bin/env node

'use strict';

function readLine()
{
    return inputString[currentLine++];
}

function superDigitHelper(num)
{
    // If we only have one digit left, then that's the super digit.

    if (num < 10)
        return num;

    // Otherwise, we add up all the digits of our number and then recurse until
    // we find our super digit. To do this, we add the % 10 and then divide the
    // number / 10 to "shift" to the next digit.

    let nextSuper = 0;

    while (num > 0)
    {
        nextSuper += num % 10;
        num = Math.trunc(num / 10);
    }

    return superDigitHelper(nextSuper);
}

function superDigit(n, k)
{
    // We only need to calculate the sum of digits of the first repetition of
    // the number, since then we can simply multiply it by k, rather than
    // constructing the potentially miles long number string.

    let start = 0

    for (let ch of n)
    {
        start += parseInt(ch, 10);
    }

    return superDigitHelper(start * k);
}

function main()
{
    const firstMultipleInput = readLine().replace(/\s+$/g, '').split(' ');
    const n = firstMultipleInput[0];
    const k = parseInt(firstMultipleInput[1], 10);
    const result = superDigit(n, k);
    console.log(result);
}

// Hackerrank setup to read from STDIN and run the app.

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');
    main();
});
